src/nucleotide.rs

```diff
@@ -2,83 +2,238 @@ use std::fmt::{self, Write};
+    Y = Nucleotide::T as u8 | Nucleotide::C as u8, // pyrimidines
+    S = Nucleotide::C as u8 | Nucleotide::G as u8,
+    K = Nucleotide::T as u8 | Nucleotide::G as u8,
+
+    B = Nucleotide::T as u8 | Nucleotide::C as u8 | Nucleotide::G as u8, // not A
+    V = Nucleotide::A as u8 | Nucleotide::C as u8 | Nucleotide::G as u8, // not T
+    D = Nucleotide::A as u8 | Nucleotide::T as u8 | Nucleotide::G as u8, // not C
+    H = Nucleotide::A as u8 | Nucleotide::T as u8 | Nucleotide::C as u8, // not G
+
+    N = Nucleotide::A as u8 | Nucleotide::T as u8 | Nucleotide::C as u8 | Nucleotide::G as u8,
+}
+
+pub trait NucleotideLike:
+    Copy + Eq + Into<u8> + Into<char> + TryFrom<u8, Error = TranslationError>
+{
+    fn complement(self) -> Self;
+    fn bits(self) -> u8;
+    fn to_ascii(self) -> u8;
+    fn is_ambiguous(self) -> bool;
+}
+
+const fn ascii_to_nucleotide_table() -> [Option<NucleotideAmbiguous>; 256] {
+    // PERF: This was previously a 128-byte array, with "high" bytes eliminated
+    // by a separate check. But extending it to 256 elements makes Rust realize
+    // it's safe to index with any u8-cast-to-usize and eliminate a bounds check
+    // from the generated code.
+    let mut pack_table = [None; 256];
+
+    macro_rules! insert {
+        ($chr:literal, $variant:expr) => {
+            pack_table[$chr.to_ascii_uppercase() as usize] = Some($variant);
+            pack_table[$chr.to_ascii_lowercase() as usize] = Some($variant);
+        };
+    }
+
+    insert!(b'a', NucleotideAmbiguous::A);
+    insert!(b't', NucleotideAmbiguous::T);
+    insert!(b'c', NucleotideAmbiguous::C);
+    insert!(b'g', NucleotideAmbiguous::G);
 
+    // ambiguity codes
+    insert!(b'n', NucleotideAmbiguous::N); // A, T, C, or G (complement N)
+    insert!(b'm', NucleotideAmbiguous::M); // A or C (complement K)
+    insert!(b'r', NucleotideAmbiguous::R); // A or G (complement Y)
+    insert!(b'w', NucleotideAmbiguous::W); // A or T (complement W, itself)
+    insert!(b's', NucleotideAmbiguous::S); // C or G (complement S, itself)
+    insert!(b'y', NucleotideAmbiguous::Y); // C or T (complement R)
+    insert!(b'k', NucleotideAmbiguous::K); // G or T (complement M)
+    insert!(b'v', NucleotideAmbiguous::V); // A, C or G (complement B)
+    insert!(b'h', NucleotideAmbiguous::H); // A, C or T (complement D)
+    insert!(b'd', NucleotideAmbiguous::D); // A, G or T (complement H)
+    insert!(b'b', NucleotideAmbiguous::B); // C, G or T (complement V)
+
+    pack_table
+}
+
+const ASCII_TO_NUCLEOTIDE: [Option<NucleotideAmbiguous>; 256] = ascii_to_nucleotide_table();
+
+impl Nucleotide {
+    pub const ALL: [Self; 4] = [Self::A, Self::T, Self::C, Self::G];
     pub const PURINES: [Self; 2] = [Self::A, Self::G];
     pub const PYRIMIDINES: [Self; 2] = [Self::C, Self::T];
+}
 
-    pub const M_AMBIGUITY: [Self; 2] = [Self::A, Self::C];
-    pub const R_AMBIGUITY: [Self; 2] = Self::PURINES;
-    pub const W_AMBIGUITY: [Self; 2] = [Self::A, Self::T];
```

Comment:
removing these will break all upstream libraries that use these. It will be pretty annoying to bump the version of quickdna.

Reply:
We could keep them for backwards compatibility, but migrating should just be a matter of going from `M_AMBIGUITY` to `NucleotideAmbiguous::M::possibilities()`